addon/components/connect.js

```diff
@@ -1,98 +1,113 @@
 import Ember from 'ember';
 
 const {
+  assert,
   computed,
   defineProperty,
+  getProperties,
   inject: { service },
-  isEmpty,
   run
 } = Ember;
 
-export default (stateToComputed, dispatchToActions=() => ({})) => {
+/**
+  Returns a list of keys that have different values between the two objects.
+
+  @method changedKeys
+  @return {Array} keys that have changed
+  @private
+*/
+function changedKeys(state, newState) {
+  return Object.keys(state).filter(key => {
+    return state[key] !== newState[key];
+  });
+}
+
+/**
+  Creates a read-only computed property for accessing redux state.
+
+  @method computedReduxProperty
+  @return {Function} an Ember computed property
+  @private
+*/
+function computedReduxProperty(key, getState) {
+  return computed({
+    get: () => getState()[key],
+    set: () => assert(`Cannot set redux property "${key}".`)
+  });
+}
+
+/**
+  Return an object of attrs passed to this Component.
+
+  `Component.attrs` is an object that can look like this:
+
+    {
+      myAttr: {
+        value: 'myValue'
+      }
+    }
 
-  if (!stateToComputed) {
-    stateToComputed = () => ({});
-  }
+  Ember provides that a `get` will return the value:
 
-  return Component => {
+    this.get('myAttr') === 'myValue'
 
-    return Component.extend({
+  @method getAttrs
+  @return {Object} an object of key, value for each attr
+  @private
+*/
+function getAttrs(context) {
+  const attrKeys = Object.keys(context.attrs || {});
+  return getProperties(context, attrKeys);
+}
 
+export default (stateToComputed, dispatchToActions) => {
+  return Component => {
+    return Component.extend({
       redux: service(),
 
       init() {
         const redux = this.get('redux');
 
-        let props = stateToComputed.call(this, redux.getState(), this.getAttrs());
+        if (stateToComputed) {
+          const getState = () => stateToComputed.call(this, redux.getState(), getAttrs(this));
 
-        Object.keys(props).forEach(name => {
-          defineProperty(this, name, computed(() =>
-            stateToComputed.call(this, redux.getState(), this.getAttrs())[name]
-          ).property().readOnly());
-        });
+          let state = getState();
```

Comment:
Yes, this is all the value here. I thought through the edge cases I can think of (particularly with observers, and computed properties based on redux properties), and cannot come up with any scenarios that might break. Given the synchronous nature of observers, and the fact that the existing properties are cached until `notifyPropertyChange` anyway, the behavior is almost identical.